tests/rules/test_script_failed.py

```diff
@@ -0,0 +1,168 @@
+        schema = watcher.rules.ScriptFailed.get_schema()
+        assert schema is not None
+        index = 1
+        config = self.make_config(index=1)
+        desired_rule_name = f"ScriptFailed.ScriptQueue:{index}"
+
+        rule = watcher.rules.ScriptFailed(config=config)
+        assert rule.name == desired_rule_name
+        assert isinstance(rule.alarm, watcher.Alarm)
+        assert rule.alarm.name == rule.name
+        assert rule.alarm.nominal
+        assert len(rule.remote_info_list) == 1
+        remote_info = rule.remote_info_list[0]
+        assert remote_info.name == "ScriptQueue"
+        assert remote_info.index == index
+        assert rule.name in repr(rule)
+        assert "ScriptFailed" in repr(rule)
+
+    async def test_call(self):
+        index = 1
+
+        watcher_config_dict = yaml.safe_load(
+            f"""
+            disabled_sal_components: []
+            auto_acknowledge_delay: 3600
+            auto_unacknowledge_delay: 3600
+            rules:
+            - classname: ScriptFailed
+              configs:
+              - index: {index}
+            escalation: []
+            """
+        )
+        watcher_config = types.SimpleNamespace(**watcher_config_dict)
+
+        async with salobj.Controller(name="ScriptQueue", index=index) as script_queue:
+            async with watcher.Model(
+                domain=script_queue.domain, config=watcher_config
+            ) as model:
+                model.enable()
+
+                assert len(model.rules) == 1
+                rule_name = f"ScriptFailed.ScriptQueue:{index}"
+                rule = model.rules[rule_name]
+                rule.alarm.init_severity_queue()
+
+                for (
+                    queue_enabled,
+                    queue_running,
+                    script_sal_index,
+                    process_state,
+                    script_state,
+                    expected_severity_1,
+                    expected_severity_2,
+                ) in [
+                    (
+                        True,
+                        True,
+                        1000,
+                        ScriptProcessState.RUNNING,
+                        ScriptState.RUNNING,
+                        AlarmSeverity.NONE,
+                        AlarmSeverity.NONE,
+                    ),
+                    (
+                        True,
+                        True,
+                        1001,
+                        ScriptProcessState.DONE,
+                        ScriptState.DONE,
+                        AlarmSeverity.NONE,
+                        AlarmSeverity.NONE,
+                    ),
+                    (
+                        True,
+                        False,
+                        1002,
+                        ScriptProcessState.DONE,
+                        ScriptState.FAILED,
+                        AlarmSeverity.NONE,
+                        AlarmSeverity.WARNING,
+                    ),
```

Comment:
This works, but I find it difficult to determine which parameters are set to which values. Consider making a dataclass to hold these test parameters and making a list of those. It adds some duplication of the parameter names, but makes it clearer which values are set to what:
```
@dataclasses.dataclass
class TestParams:
    queue_enabled: bool
    ...
    expected_severity_1: AlarmSeverity

for params in (
    TestParams(
         queue_enabled=True,
         queue_running=True,
         ...,
         expected_severity_2=AlarmSeverity.NONE
     ),
    ...
):
```

Reply:
Another option, which would allow you to publish a script state of FAILED for the non-current script (which your rule should ignore) is to just publish the events. More duplication, but no issue of 2 events per test data. I admit that only the current script should ever be reported as FAILED (except if it fails configuration, which is a different state in the soon-to-be-released version).

Or make your test params be for one event or the other, and publish just the one event for each set of parameters.